File: python/get_groups.py
@make(inputs=[0], outputs=[1], handler=Output.STORE)
def get_groups(ctx, show_special_groups):
    """
    Get a listing of the all project's collections

    Parameters
    ----------
    ctx : Context
        Combined type of a callback and rei struct.
    show_special_groups: str
        Project absolute path

    Returns
    -------
    list
        a json list of collections objects
    """
    # Initialize the groups array
    groups = []

    for item in row_iterator("USER_NAME, USER_ID", "USER_TYPE = 'rodsgroup'", AS_LIST, ctx.callback):
        group_name = item[0]
        group_id = item[1]
        display_name = item[0]
        description = ""

        for meta in row_iterator("META_USER_ATTR_NAME, META_USER_ATTR_VALUE, USER_GROUP_ID, USER_GROUP_NAME",
                                 "USER_TYPE = 'rodsgroup' and USER_GROUP_ID = '{}'".format(group_id),
                                 AS_LIST, ctx.callback):
            if "displayName" == meta[0]:
                display_name = meta[1]
            elif "description" == meta[0]:
                description = meta[1]

        group_object = {
            "name": group_name,
            "groupId": group_id,
            "displayName": display_name,
            "description": description
        }
        if show_special_groups == "false":
            if group_name not in ["public", "rodsadmin", "DH-ingest", "DH-project-admins"]:
                groups.append(group_object)
        else:
            groups.append(group_object)

    return groups

File: native_irods_ruleset/misc/getGroups.r
# Call with
#
# irule -r irods_rule_engine_plugin-irods_rule_language-instance -F /rules/native_irods_ruleset/misc/getGroups.r "*showSpecialGroups='true'"

irule_dummy() {
    IRULE_getGroups(*showSpecialGroups, *result);
    writeLine("stdout", *result);
}

IRULE_getGroups(*showSpecialGroups, *result) {
    *groups = '[]';

    foreach ( *Row in select USER_NAME, USER_ID where USER_TYPE = 'rodsgroup') {
        *objectName = *Row.USER_NAME;
        *objectID = *Row.USER_ID;
        *displayName = *objectName;
        *description = "";

        foreach (*av in SELECT META_USER_ATTR_NAME, META_USER_ATTR_VALUE, USER_GROUP_ID, USER_GROUP_NAME where USER_TYPE = 'rodsgroup' and USER_GROUP_ID = *objectID ) {
           if( "displayName" == *av.META_USER_ATTR_NAME ) {
              *displayName = *av.META_USER_ATTR_VALUE;
           }
           else if( "description" == *av.META_USER_ATTR_NAME ) {
              *description = *av.META_USER_ATTR_VALUE;
           } 
        }

        *groupObject = '{ "userName" : "*objectName", "userId" : "*objectID", "displayName" : "*displayName", "description" : "*description" }';

        if ( str(*showSpecialGroups) == "false" ) {
            if ( *Row.USER_NAME != "public" &&  *Row.USER_NAME != "rodsadmin" && *Row.USER_NAME != "DH-ingest" && *Row.USER_NAME != "DH-project-admins" ) {
                json_arrayops_add(*groups, *groupObject, "");
            }
         } else {
            json_arrayops_add(*groups, *groupObject, "");
         }
    }

    *result = *groups;
}

INPUT *showSpecialGroups='true'
OUTPUT ruleExecOut
